esmvaltool/diag_scripts/weighting/climwip/calibrate_sigmas.py

```diff
@@ -0,0 +1,374 @@
+
+    Parameters
+    ----------
+    target : array_like, shape (N,)
+        See calculate_percentiles for more information.
+    weights_matrix : array_like, shape (N, N)
+        See calculate_percentiles for more information.
+    performance_sigma : float
+        The performance sigma value used to calculate the weights.
+
+    Returns
+    -------
+    cost_function_value : float
+        Value of the cost function based on the given performance sigma value.
+        The cost function is a discontinuous function distinguishing two cases:
+        * overconfident weighting (ratio_difference < 0.):
+            The value is positive and decreases with increasing sigma
+        * not overconfident (reliable) weighting (ratio_difference >= 0.):
+            The value is negative and increases with increasing sigma
+
+    WARNING
+    -------
+    - It is highly recommended to visually inspect the graphical output of this
+    process to ensure the optimisation worked as intended.
+
+    - Using cases where the largest possible sigma value was pickt is not
+    recommended. This is normally a sign that
+        * not enough models where used
+        * the diagnostics where not well chose with respect to the target
+
+    Additional information
+    ----------------------
+    After evaluating all possible sigma values the sigma which leads to the
+    smallest cost function will be selected. Two different cases need to be
+    disdinguished:
+    * the cost function is negative for at least one sigma:
+        For this case the cost function increases linearly with sigma,
+        therefore the smallest possible sigma where this is true will be
+        selected.
+    * the cost function is non-negative:
+        For this case the cost function decreases linearly with sigma and
+        with decreasing absolute value ratio_difference. This will normally
+        lead to the largest possible sigma value being selected.
+    """
+    percentiles = PERCENTILES
+    inside_ratio_reference = percentiles[1] - percentiles[0]
+
+    # calculate the equally weighted case once as baseline
+    if len(confidence_test_values['baseline']) == 0:
+        percentiles_spread, inside_ratio = calculate_percentiles(
+            target, weights_matrix, percentiles, weighted=False)
+        confidence_test_values['baseline']['percentile_spread'] = (
+            percentiles_spread)
+        confidence_test_values['baseline']['inside_ratio'] = inside_ratio
+
+    percentiles_spread, inside_ratio = calculate_percentiles(
+        target, weights_matrix, percentiles)
+    confidence_test_values[performance_sigma] = {
+        'percentile_spread': percentiles_spread,
+        'inside_ratio': inside_ratio}
+
+    ratio_difference = inside_ratio - inside_ratio_reference
+
+    if ratio_difference < 0:
+        return np.abs(ratio_difference) + SIGMA_RANGE[1] - performance_sigma
+    return performance_sigma - SIGMA_RANGE[1]
```

Comment:
This block looks still a bit opaque to me. Can we make it more readable, e.g. by something like this:

```suggestion
    overconfident = inside_ratio - inside_ratio_reference < 0
    best_compromise = np.abs(ratio_difference) + SIGMA_RANGE[1] - performance_sigma
    reliable_value = performance_sigma - SIGMA_RANGE[1]
    cost =  best_compromise if overconfident else reliable_value
    return cost
```